engine/consensus/ingestion/engine.go

```diff
@@ -281,3 +254,93 @@ func (e *Engine) validateExpiry(guarantee *flow.CollectionGuarantee) error {
 
 	return nil
 }
+
+// validateGuarantors validates that the guarantors of a collection are valid,
+// in that they are all from the same cluster and that cluster is allowed to
+// produce the given collection w.r.t. the guarantee's reference block.
+func (e *Engine) validateGuarantors(guarantee *flow.CollectionGuarantee) error {
+
+	guarantors := guarantee.SignerIDs
+
+	if len(guarantors) == 0 {
+		return engine.NewInvalidInputError("invalid collection guarantee with no guarantors")
+	}
+
+	// get the clusters to assign the guarantee and check if the guarantor is part of it
+	snapshot := e.state.AtBlockID(guarantee.ReferenceBlockID)
+	clusters, err := snapshot.Epochs().Current().Clustering()
+	if err != nil {
+		return fmt.Errorf("could not get clusters: %w", err)
+	}
+	cluster, _, ok := clusters.ByNodeID(guarantors[0])
+	if !ok {
+		return engine.NewInvalidInputErrorf("guarantor (id=%s) does not exist in any cluster", guarantors[0])
+	}
+
+	// NOTE: Eventually we should check the signatures, ensure a quorum of the
+	// cluster, and ensure HotStuff finalization rules. Likely a cluster-specific
+	// version of the follower will be a good fit for this. For now, collection
+	// nodes independently decide when a collection is finalized and we only check
+	// that the guarantors are all from the same cluster.
+
+	// ensure the guarantors are from the same cluster
+	clusterLookup := cluster.Lookup()
+	for _, guarantorID := range guarantors {
+		_, exists := clusterLookup[guarantorID]
+		if !exists {
+			return engine.NewInvalidInputError("inconsistent guarantors from different clusters")
+		}
+	}
+
+	return nil
+}
+
+// validateOrigin validates that the message has a valid sender (origin). We
+// allow Guarantee messages either from collection nodes or from consensus nodes.
+// Collection nodes must be valid in the identity table w.r.t. the reference
+// block of the collection; consensus nodes must be valid w.r.t. the latest
+// finalized block.
+//
+// Returns the origin identity as validated if validation passes.
+func (e *Engine) validateOrigin(originID flow.Identifier, guarantee *flow.CollectionGuarantee) (*flow.Identity, error) {
+
+	// get the origin identity w.r.t. the collection reference block
+	origin, err := e.state.AtBlockID(guarantee.ReferenceBlockID).Identity(originID)
+	if err != nil && !protocol.IsIdentityNotFound(err) {
+		return nil, engine.NewInvalidInputErrorf("could not get origin (id=%x) identity w.r.t. reference block : %w", originID, err)
+	}
```

Comment:
I don't think _any_ possible error (excluding `IdentityNotFound`) is necessarily an indicator of an invalid input error. I think we should carefully handle the errors we are expecting and escalate all others as symptoms of node-internal problems.